aptos-move/framework/src/natives/cryptography/algebra/arithmetics/scalar_mul.rs

```diff
@@ -0,0 +1,223 @@
+// Copyright © Aptos Foundation
+
+use crate::{
+    abort_unless_feature_flag_enabled,
+    natives::{
+        cryptography::algebra::{
+            abort_invariant_violated, gas::GasParameters, AlgebraContext, Structure,
+            MOVE_ABORT_CODE_INPUT_VECTOR_SIZES_NOT_MATCHING, MOVE_ABORT_CODE_NOT_IMPLEMENTED,
+        },
+        helpers::{log2_ceil, SafeNativeContext, SafeNativeError, SafeNativeResult},
+    },
+    safe_borrow_element, safely_pop_arg, store_element, structure_from_ty_arg,
+};
+use aptos_types::on_chain_config::FeatureFlag;
+use ark_ec::{CurveGroup, Group};
+use ark_ff::Field;
+use move_core_types::gas_algebra::{InternalGas, InternalGasPerArg, NumArgs};
+use move_vm_types::{loaded_data::runtime_types::Type, values::Value};
+use smallvec::{smallvec, SmallVec};
+use std::{collections::VecDeque, rc::Rc};
+
+fn feature_flag_of_group_scalar_mul(
+    group_opt: Option<Structure>,
+    scalar_field_opt: Option<Structure>,
+) -> Option<FeatureFlag> {
+    match (group_opt, scalar_field_opt) {
+        (Some(Structure::BLS12381G1Affine), Some(Structure::BLS12381Fr))
+        | (Some(Structure::BLS12381G2Affine), Some(Structure::BLS12381Fr))
+        | (Some(Structure::BLS12381Gt), Some(Structure::BLS12381Fr)) => {
+            Some(FeatureFlag::BLS12_381_STRUCTURES)
+        },
+        _ => None,
+    }
+}
+
+macro_rules! abort_unless_group_scalar_mul_enabled {
+    ($context:ident, $group_opt:expr, $scalar_field_opt:expr) => {
+        let flag_opt = feature_flag_of_group_scalar_mul($group_opt, $scalar_field_opt);
+        abort_unless_feature_flag_enabled!($context, flag_opt);
+    };
+}
+
+macro_rules! ark_scalar_mul_internal {
+    ($context:expr, $args:ident, $group_typ:ty, $scalar_typ:ty, $op:ident, $gas:expr) => {{
+        let scalar_handle = safely_pop_arg!($args, u64) as usize;
+        let element_handle = safely_pop_arg!($args, u64) as usize;
+        safe_borrow_element!($context, element_handle, $group_typ, element_ptr, element);
+        safe_borrow_element!($context, scalar_handle, $scalar_typ, scalar_ptr, scalar);
+        let scalar_bigint: ark_ff::BigInteger256 = (*scalar).into();
+        $context.charge($gas)?;
+        let new_element = element.$op(scalar_bigint);
+        let new_handle = store_element!($context, new_element);
+        Ok(smallvec![Value::u64(new_handle as u64)])
+    }};
+}
+
+fn ark_msm_window_size(num_entries: usize) -> usize {
```

Comment:
/// WARNING: Be careful with the unwrap() below, if you modify this if statement.

Reply:
Fixed.